PG-1858 Document backing up with WAL with encrypt enabled

Andriciuc opened the pull request:
- add new topic called # Backing up with WAL encryption enabled

Andriciuc committed:
PG-1858 Document backing up with WAL with encrypt enabled

Andriciuc commented:
Hello @dAdAbird, I created a new topic in the TDE operations regarding this ticket, should we also add a quick ## Example procedure to include users who are less familiar with backing up? 

Something like:

# On the source server: copy the pg_tde directory 
code
# On the backup server: take the base backup
code

<img width="160" height="187" alt="backup" src="https://github.com/user-attachments/assets/3336b6a8-61e6-4794-80f4-3990870e874e" />

dAdAbird reviewed (commented):
1. Quite a bit changed, see "How to document" section here:https://perconadev.atlassian.net/browse/PG-1857

2. We should clarify that the user has to not only copy the pg_tde dir, but all files referenced in providers config (`/path_to/server_certificate.pem` etc). Otherwise the server key won't be accessible.

Andriciuc committed:
Update the Features topic buttons for better clarity (#508)

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into backup-wal-enabled

Andriciuc committed:
update the backup topic with fresh information

Andriciuc committed:
Merge branch 'release-17.5.3' into backup-wal-enabled

Andriciuc committed:
Merge branch 'release-17.5.3' into backup-wal-enabled

dAdAbird commented on contrib/pg_tde/documentation/docs/how-to/backup-wal-enabled.md:
````diff
@@ -0,0 +1,19 @@
+2. Run:
+
+    ```bash
+    pg_basebackup -X stream -F p -E
````
1. The `-D /path/to/backup` argument is mandatory for the `-F p`. `/path/to/backup` is this example would be the to which user has to copy `pg_tde`.
2.  `-X stream` and `-F p` are defaults so they can be omitted. Usually you'd put the `-X` of `-F` flag if you want to set non-default values.

So I'd leave only 
```bash
pg_basebackup -D /path/to/backup -E
```

And we should a node that -E works/makes sense only with the wal method "stream" (hence it doesn't make sense and won't work with `-X none` or `-X fetch`). And that -E doesn't work with the tar output format ('-F t').

Andriciuc replied:
I have rewritten the run command with the above changes, also made sure to explain to the user in the notes the specific limitations for tar output format and wal method stream option:

2. Run:

    ```bash
    pg_basebackup -D /path/to/backup -E
    ```

    Where:

    - `-D /path/to/backup` specifies the backup location where you have to copy `pg_tde`
    - `-E` (or `--encrypt-wal`) enables WAL encryption and validates that the copied `pg_tde` and provider files are present and that the server key is accessible (required)

!!! note
    - The `-E` flag only works with the `--wal-method=stream` option (default). It is not compatible with `-X none` or `-X fetch`.
    - The `-E` flag is only supported with the plain output format (`-F p`). It cannot be used with the tar output format (`-F t`).

Andriciuc replied:
Updated the command and rewrote the note to reflect proper flag explanations.

dAdAbird commented on contrib/pg_tde/documentation/docs/how-to/backup-wal-enabled.md:
```diff
@@ -0,0 +1,19 @@
+
+To create a backup with WAL encryption enabled:
+
+1. Copy the `pg_tde` directory, including the `wal_keys` and `1664_providers` files, and any external files referenced by your providers configuration (for example, certificate or key files).
```
It's not clear where to copy the `pg_tde` dir. And the external files for that matter

Andriciuc replied:
Expanded this intro with an example path and more detail on what `pg_basebackup -E` validates.

dAdAbird commented:
We also need to document the restore for the backup taken with `-E`. That the server key and all the external files referenced by providers should be accessible on the restored server

Andriciuc committed:
add two suptopics for other wal methods and restore backup created with wal encrypt

- replace pg_basebackup command with proper flags
- add more details to intro about where the user should copy the data to and from

dAdAbird commented on contrib/pg_tde/documentation/docs/how-to/backup-wal-enabled.md:
```diff
@@ -0,0 +1,44 @@
+
+When you want to restore a backup created with `pg_basebackup -E`:
+
+1. Ensure that the `pg_tde` directory from the backup you have done above is placed in the restored server’s data directory.
```
No need to do anything special pg_tde. It's a part of the backup now and should be restored to the destination as any other file.

Andriciuc replied:
removed, thank you for clarifying!

dAdAbird commented on contrib/pg_tde/documentation/docs/how-to/backup-wal-enabled.md:
```diff
@@ -0,0 +1,44 @@
+
+## Other WAL methods
+
+The `-X none` and `-X fetch` options are not affected by WAL encryption and both work with encrypted WAL without requiring any additional flags.  
```
`not affected by WAL encryption and both work with encrypted WAL` sounds a bit confusing to me. Not sure how to rephrase it though. Maybe just leave "The `-X none` and `-X fetch` options work with encrypted WAL without requiring any additional flags." But `-X none` ins't concerned with WAL at all (no WAL in backup)...

Andriciuc replied:
I see the point and this is very true! Since we mention fetch, none and stream, lets keep both mentions just to cover our bases in case users ask "what about none?":

The `-X fetch` option works with encrypted WAL without requiring any additional flags.  
The `-X none` option excludes WAL from the backup and is unaffected by WAL encryption.

Also, question, we use a mix of longform like `--wal-method=stream` and shortform like `-X stream` for all 3 flags. I would suggest we just use the short form for all of them, thoughts?

dAdAbird replied:
> The -X fetch option works with encrypted WAL without requiring any additional flags.
> The -X none option excludes WAL from the backup and is unaffected by WAL encryption.

Looks good!


> Also, question, we use a mix of longform like --wal-method=stream and shortform like -X stream for all 3 flags. I would suggest we just use the short form for all of them, thoughts?

Yes, let's stick to one form - `-X ...`, this is what [pg_waldump documentation](https://www.postgresql.org/docs/current/app-pgbasebackup.html) uses as well.

Andriciuc replied:
Done!

Andriciuc committed:
reword to short form option flags

- improve restore backup steps
- split other wal methods into two separate description paragraphs

Andriciuc committed:
Merge branch 'release-17.5.3' into backup-wal-enabled

dAdAbird reviewed (approved):
Looks good!
